### ISSUE TITLE
gradient of posterior mean w.r.t. x

### ISSUE THREAD

**Issue by shalijiang** [pytorch/botorch#264]
Is there a way to get the gradient of the posterior mean w.r.t. x?
For example, 

`
  posterior = model.posterior(x)

  x_grad = torch.autograd.grad(posterior.mean, x, create_graph=True)[0]
`

But seems currently the posterior computation is harded coded to be eval mode, so posterior.mean does not have a grad_fn

Is there a way around this?

**Comment by Balandat** [comment]
This should work out of the box. Did you make sure that your input `x` has `requires_grad=True`?

```
x = torch.rand(2, 1, requires_grad=True)
pmean = gp.posterior(x).mean
```

While looking into this I found a pretty nasty bug though, namely that the gradient when computing it at a single point (e.g. x is `1 x d`) is non-deterministic and all over the place - I'll open up a separate issue for that.

**Comment by shalijiang** [comment]
I did add 
`x.requires_grad = True`,
doesn't seem to work. 
I also did exactly as your two lines of code, I don't see pmean has a grad_fn. 
Did you run it? could you copy the runnable code here?
Thanks!

**Comment by Balandat** [comment]
I'm going to close this out - feel free to re-open if the above didn't solve your issue.

Btw, the bug I mentioned happens to be a core pytorch bug: https://github.com/pytorch/pytorch/issues/26076
